### PR TITLE
Task 4 form

### PR DESCRIPTION
**Pull request by pwalawko**

### CODE REVIEW

File: task3_form/FlaskApp/app.py [hunk 1]
```diff
@@ -8,12 +9,13 @@
 
 @app.route("/", methods=['GET', 'POST'])
 def show_page():
-    if request.method == 'POST':
-        person = request.form['person']
-        food = request.form['food']
+    form = OrderForm(request.form, person='')
+    if request.method == 'POST' and form.validate():
```
Comment (kurazu) [inline task3_form/FlaskApp/app.py]: 👍

File: task3_form/FlaskApp/forms.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+from wtforms import Form, StringField, validators, ValidationError
+
+
+CHARACTERS = ('znak', 'znaki', 'znaków')
+req_field_msg = "Nie zapominaj o tym polu."
+
+
+def plural_form(n):
```
Comment (kurazu) [inline task3_form/FlaskApp/forms.py]: 👍 Good you are using what you learned earlier :)

File: task3_form/FlaskApp/forms.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        else:
+            plural = 2
+
+    return plural
+
+
+def length(min=-1, max=-1):
+    max_char_written = CHARACTERS[plural_form(max)]
+    min_char_written = CHARACTERS[plural_form(min)]
+    message_max = 'Nie przesadzasz z tą ilością? \
```
Comment (kurazu) [inline task3_form/FlaskApp/forms.py]: Please use [formatted string literals](https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep498) for text formatting.

File: task3_form/FlaskApp/forms.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        Wpisz mniej niż %d %s.' % (max, max_char_written)
+    message_min = 'Co tak skromnie? \
+        Wpisz więcej niż %d %s.' % (min, min_char_written)
+
+    def _length(form, field):
+        le = field.data and len(field.data) or 0
```
Comment (kurazu) [inline task3_form/FlaskApp/forms.py]: It's recommended to insert additional parenthesis where the order of operations is not immediately clear (here it's determined by the precedence of operators).

File: task3_form/FlaskApp/forms.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+def length(min=-1, max=-1):
+    max_char_written = CHARACTERS[plural_form(max)]
+    min_char_written = CHARACTERS[plural_form(min)]
+    message_max = 'Nie przesadzasz z tą ilością? \
+        Wpisz mniej niż %d %s.' % (max, max_char_written)
+    message_min = 'Co tak skromnie? \
+        Wpisz więcej niż %d %s.' % (min, min_char_written)
+
+    def _length(form, field):
```
Comment (kurazu) [inline task3_form/FlaskApp/forms.py]: 👍 functions returning functions

File: task3_form/FlaskApp/forms.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    else:
+        if n % 10 >= 2 and n % 10 <= 4 and (n % 100 < 10 or n % 100 >= 20):
+            plural = 1
+        else:
+            plural = 2
+
+    return plural
+
+
+def length(min=-1, max=-1):
```
Comment (kurazu) [inline task3_form/FlaskApp/forms.py]: I'd argue that `None` would be a better default value.

File: task3_form/FlaskApp/forms.py (hunk 3 above)
Comment (pwalawko) [reply task3_form/FlaskApp/forms.py]: done

File: task3_form/FlaskApp/forms.py (hunk 4 above)
Comment (pwalawko) [reply task3_form/FlaskApp/forms.py]: done

File: task3_form/FlaskApp/forms.py (hunk 6 above)
Comment (pwalawko) [reply task3_form/FlaskApp/forms.py]: done

File: task3_form/FlaskApp/forms.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            plural = 2
+
+    return plural
+
+
+def length(min_le=None, max_le=None):
+
+    def _length(form, field):
+        le = (field.data and len(field.data)) or 0
+        if min_le is not None and le < min_le:
```
Comment (kurazu) [inline task3_form/FlaskApp/forms.py]: :+1:

File: task3_form/FlaskApp/app.py [hunk 8]
```diff
@@ -1,19 +1,67 @@
...
+        return 0
+    if n % 10 >= 2 and n % 10 <= 4 and (n % 100 < 10 or n % 100 >= 20):
+        return 1
+    return 2
+
+
+def length_validation(field, req=False, min_le=None, max_le=None):
+    field_length = (field and len(field)) or 0
+    if field_length == 0:
+        if req is True:
```
Comment (kurazu) [inline task3_form/FlaskApp/app.py]: Most ppl would put is just as ```if req:```

File: task3_form/FlaskApp/app.py [hunk 9]
```diff
@@ -1,19 +1,67 @@
...
+                person=person,
+                food=food,
+                person_val_msg=person_val_msg,
+                food_val_msg=food_val_msg,
+                person_is_valid=person_is_valid,
+                food_is_valid=food_is_valid
+            )
+
         orders.append([person, food])
-        return redirect("/")
```
Comment (kurazu) [inline task3_form/FlaskApp/app.py]: This is a grievous error. You got rid of the post-redirect-get pattern.

Please follow the steps below to see why this is bad:
1. Open the app
2. Fill in (correctly) the form
3. Submit
4. The order is added and rendered on the page
5. Meantime Refał does the same steps from a different browser.
6. Now you want to see both yours and his orderes, so you hit F5.
7. The browser displays a popup about form resubmission (⚠️ )
8. You confirm the popup, the page reloads
9. Turns  out your order was added for the second time! (👎 )

Post-redirect-get is essential to making sure that the action of refreshing the page will not side-effects.
Please restore post-redirect-get.

File: task3_form/FlaskApp/templates/index.html [hunk 10]
```diff
@@ -18,18 +23,28 @@ <h1>ZAMÓWIENIA</h1>
         <div id="add_order">
             <h3>Dodawanie zamówienia</h3>
             <form id="add_order" method="POST">
-                <div>
+                <div class="form-group">
```
Comment (kurazu) [inline task3_form/FlaskApp/templates/index.html]: 👍 Now the form looks pretty!

**Pull request merged**